### ISSUE TITLE
use case: using zig cc to cross compile a cgo program for windows

### ISSUE THREAD

**Issue by andrewrk** [golang/go#43886]
Hello, any Go developers interested in working with Zig developers on this use case?

Upstream issue: https://github.com/ziglang/zig/issues/7874
Tested with go 1.15.5.

main.go

```go
package main

//int Add(int a, int b){
//    return a+b;
//}
import "C"
import "fmt"

func main() {
	a := C.int(10)
	b := C.int(20)
	c := C.Add(a, b)
	fmt.Println(c) // 30
}
```

`CGO_ENABLED=1 GOOS=windows GOARCH=amd64 CC="zcc" CXX="zc++" go build main.go`

zcc
```sh
#!/bin/sh
zig cc -target x86_64-windows-gnu $@
```

zc++
```
#!/bin/sh
zig c++ -target x86_64-windows-gnu $@
```

output:

```
# command-line-arguments
warning: unsupported linker arg: --compress-debug-sections=zlib-gnu
```

It produces `main.exe` but when run it segfaults in `runtime.rt0_go()` (called from main()). Expected output is to print "30" to the console when run in cmd.exe.

Happy to chat on IRC or Discord if anyone has some ideas to try.

**Comment by AlexRouSg** [comment]
Duplicate of #43078

**Comment by AlexRouSg** [comment]
@andrewrk does anything change if you add the `-buildmode=exe` flag to `go build`?

**Comment by andrewrk** [comment]
no- still get a segfault at:

```
0x000000014013b53c runtime.rt0_go+0x5c in main: movb	$0x1,0xfffffffffffccee7(%rip)
```

was go/cgo expecting the linker to do something that perhaps didn't get done here?

**Comment by AlexRouSg** [comment]
@ianlancetaylor would be the best person to talk to about this but I'm not too sure if this should be discussed here cause the issue tracker is normally for bugs and zig isn't exactly the intended/expected compiler/linker.

**Comment by andrewrk** [comment]
hmm I think this may be an issue with cgo though; is that in scope for this issue tracker?

**Comment by dominikh** [comment]
Let's keep this issue open until we know if this is a bug in zig or in Go.

**Comment by mdempsky** [comment]
I don't have any familiarity with using zig. Can you give instructions on how to easily reproduce this issue from a clean Linux install?

A docker image would be ideal.

**Comment by Sobeston** [comment]
Reproduced with go 1.15.6 linux/amd64, full output here: https://pastebin.com/BmVtawS1

@mdempsky Assuming you have go installed already, all you should need is a master copy of zig. For you this probably means getting a binary from https://ziglang.org/download/ and adding it to your path (shouldn't need any more setup than this). Personally I get my version from the "zig-git" aur package.

**Comment by Sobeston** [comment]
Also tested on go 1.15.7 linux/amd64; same problem https://pastebin.com/wHw49gWf.

**Comment by LaPingvino** [comment]
I thought maybe filtering out the zlib-gnu invocation from Go might fix it, which I did with using this for the shell wrapper:

```bash
#!/bin/bash
INVOKE=${@/-Wl,--compress-debug-sections=zlib-gnu/}
echo $INVOKE
zig cc -target x86_64-windows-gnu $INVOKE
```

(set invoke to $@ to see the full invocation that CGo does)

and that does seem to solve the error, but I'm not sure if the resulting binary is as expected, adding it here, cannot test because compiled on Linux ARM64: [zigandgotest.zip](https://github.com/golang/go/files/5979674/zigandgotest.zip)

**Comment by LaPingvino** [comment]
Update: found a Windows machine to test on, running that binary doesn't give any errors but doesn't produce the desired result (fyne hello world) either.

**Comment by zx2c4** [comment]
Likely fixed by https://go-review.googlesource.com/c/go/+/291630

**Comment by LaPingvino** [comment]
Context to @zx2c4's reference (correct me if I'm wrong...): LLVM builds seem to have been broken for Windows with CGo before, but it wasn't really an issue until trying to get ARM64 on Windows to work. Fixes are applied now to get Windows/ARM64 working, and those should fix Windows builds across the board.

**Comment by zx2c4** [comment]
Those fixes haven't been applied yet. I'm hoping @rsc will submit them with the rest of the windows/arm64 series in the near future.

**Comment by LaPingvino** [comment]
Yeah, the patch you refer to is still stuck in trybot-hell atm xD. If you have a branch with these fixes applied I would love to test that for the purpose of this bug.

**Comment by zx2c4** [comment]
Here's what I'm building with at the moment: https://download.wireguard.com/windows-toolchain/distfiles/go1.16-2021-02-18.zip It's 1.16 plus a bunch of extras. Cgo works with Clang but I haven't tried zig's wrapper yet. Let me know how it goes.

**Comment by LaPingvino** [comment]
```bash
$ CC="zcc" CGO_ENABLED=1 GOOS=windows GOARCH=amd64 go build
# fyne-hello
warning: unsupported linker arg: --compress-debug-sections=zlib-gnu
```
[zigandgotest.zip](https://github.com/golang/go/files/6006376/zigandgotest.zip)

**Comment by LaPingvino** [comment]
Tested with wine on my amd64 system: works. Might need to file the unsupported linker arg thing separately.

**Comment by LaPingvino** [comment]
The argument order patch should fix the zlib-gnu warning, @zx2c4 said to check out that patch and with a bit of luck I will be able to test that too.

**Comment by LaPingvino** [comment]
@zx2c4 as we found that your version works, maybe link/mention the relevant bugs here so Github can help us keep track if we can close this bug?

**Comment by LaPingvino** [comment]
> The argument order patch should fix the zlib-gnu warning, @zx2c4 said to check out that patch and with a bit of luck I will be able to test that too.

Doesn't seem to be an order patch yet? Current master seems to fix issue otherwise 👍  (while still needing the workaround)

**Comment by qmuntal** [comment]
[CL 291630](https://go-review.googlesource.com/c/go/+/291630) fixed this issue. Closing.
